test/replication-luatest/gh_5272_recovery_triggers_test.lua

```diff
@@ -0,0 +1,85 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('gh-5272')
+
+local run_before_cfg = [[
+    box.ctl.on_recovery_state(function(state)
+        require('log').info('State ' .. state .. ' reached')
+    end)
```

Comment:
It might be easier to inspect the states not via logs but via a Lua table, like that:
```Diff
@@ -5,8 +5,9 @@ local server = require('test.luatest_helpers.server')
 local g = t.group('gh-5272')
 
 local run_before_cfg = [[
+    rawset(_G, 'test_states', {})
     box.ctl.on_recovery_state(function(state)
-        require('log').info('State ' .. state .. ' reached')
+        table.insert(test_states, state)
     end)
 ]]
```
Then in the test you can check `test_states` array without reading any logs. Up to you.

Reply:
Good point, done:
```diff
diff --git a/test/replication-luatest/recovery_triggers_test.lua b/test/replication-luatest/recovery_triggers_test.lua
index dc0c88574..d15d00816 100644
--- a/test/replication-luatest/recovery_triggers_test.lua
+++ b/test/replication-luatest/recovery_triggers_test.lua
@@ -5,8 +5,9 @@ local server = require('test.luatest_helpers.server')
 local g = t.group('gh-5272')
 
 local run_before_cfg = [[
+    rawset(_G, 'test_states', {})
     box.ctl.on_recovery_state(function(state)
-        require('log').info('State ' .. state .. ' reached')
+        table.insert(test_states, state)
     end)
 ]]
 
@@ -29,16 +30,27 @@ g.after_all(function(cg)
     cg.cluster:drop()
 end)
 
+local function is_state_reached(server, state)
+    return server:exec(function(state)
+        for _, s in pairs(test_states) do
+            if s == state then
+                return true
+            end
+        end
+        return false
+    end, {state})
+end
+
 local function wait_state(server, state)
     t.helpers.retrying({}, function()
-        t.assert(server:grep_log('State '.. state .. ' reached'),
+        t.assert(is_state_reached(server, state),
                  'State ' .. state .. ' triggers fired')
     end)
 end
 
 local function assert_state_not_reached(server, state)
-    t.assert_equals(server:grep_log('State ' .. state .. ' reached'), nil,
-                    'State ' .. state .. ' not reached')
+    t.assert(not is_state_reached(server, state),
+             'State ' .. state .. ' not reached')
 end
 
 local function wait_usual_state_order(server)
```